glyphs-reader/src/font.rs

```diff
@@ -333,24 +335,84 @@ impl FromPlist for LayerAttributes {
             let key: String = tokenizer.parse()?;
             tokenizer.eat(b'=')?;
             match key.as_str() {
-                "coordinates" => {
-                    coordinates = tokenizer.parse()?;
-                }
+                "coordinates" => coordinates = tokenizer.parse()?,
+                "color" => color = tokenizer.parse::<i64>()? == 1,
                 // skip unsupported attributes for now
                 // TODO: match the others
                 _ => tokenizer.skip_rec()?,
             }
             tokenizer.eat(b';')?;
         }
 
-        Ok(LayerAttributes { coordinates })
+        Ok(LayerAttributes { coordinates, color })
+    }
+}
+
+#[derive(Clone, Default, Debug, PartialEq, Eq, Hash, FromPlist)]
+pub struct ShapeAttributes {
+    pub gradient: Gradient,
+}
+
+#[derive(Clone, Default, Debug, PartialEq, Eq, Hash, FromPlist)]
+pub struct Gradient {
+    start: Vec<OrderedFloat<f64>>,
+    end: Vec<OrderedFloat<f64>>,
+    colors: Vec<Color>,
+    #[fromplist(key = "type")]
+    style: String,
+}
+
+#[derive(Clone, Default, Debug, PartialEq, Eq, Hash)]
+pub struct Color {
+    pub r: i64,
+    pub g: i64,
+    pub b: i64,
+    pub a: i64,
+    pub n: i64, // 0/1 for start/end
+}
+
+// hand-parse because it's a list of inconsistent types
+impl FromPlist for Color {
+    fn parse(tokenizer: &mut Tokenizer<'_>) -> Result<Self, crate::plist::Error> {
+        tokenizer.eat(b'(')?;
+
+        let colors = tokenizer.parse::<Vec<i64>>()?;
+        tokenizer.eat(b',')?;
+        let n = tokenizer.parse::<i64>()?;
+        tokenizer.eat(b')')?;
+
+        if colors.len() != 4 {
+            return Err(crate::plist::Error::UnexpectedNumberOfValues {
+                value_type: "rgba values",
+                expected: 4,
+                actual: colors.len(),
+            });
+        }
+
+        Ok(Color {
+            r: colors[0],
+            g: colors[1],
+            b: colors[2],
+            a: colors[3],
+            n,
+        })
     }
 }
 
 #[derive(Debug, Clone, PartialEq, Hash)]
 pub enum Shape {
-    Path(Path),
-    Component(Component),
+    Path(Path, ShapeAttributes),
+    Component(Component, ShapeAttributes),
```

Comment:
why did you add ShapeAttributes to the enum values instead of as actual fileds of Component and Path structs?

in the glyphs format both Component and Path have attributes directly, and glyphsLib also models them like that
https://github.com/schriftgestalt/GlyphsSDK/blob/Glyphs3/GlyphsFileFormat/GlyphsFileFormatv3.md

Reply:
Agreed that this probably makes a bit more sense; it means that you can extract the inner type without having to discard the attributes, which might be useful at some point?